src/iterator/backend.rs

```diff
@@ -0,0 +1,380 @@
+        // Already registered, ignoring
+        if lock[signal as usize].is_some() {
+            return Ok(());
+        }
+
+        let waker = Arc::clone(&self.waker);
+        let action = move || {
+            waker.pending[signal as usize].store(true, Ordering::SeqCst);
+            waker.wake();
+        };
+        let id = unsafe { crate::register(signal, action) }?;
+        lock[signal as usize] = Some(id);
+        Ok(())
+    }
+
+    /// Get a reference to the read end of the self pipe
+    pub fn get_read(&self) -> &R {
+        &self.waker.read
+    }
+
+    /// Drains all data from the internal self-pipe. This method will never block.
+    fn flush(&self) {
+        const SIZE: usize = 1024;
+        let mut buff = [0u8; SIZE];
+
+        unsafe {
+            // Draining the data in the self pipe. We ignore all errors on purpose. This
+            // should not be something like closed file descriptor. It could EAGAIN, but
+            // that's OK in case we say MSG_DONTWAIT. If it's EINTR, then it's OK too,
+            // it'll only create a spurious wakeup.
+            while libc::recv(
+                self.waker.read.as_raw_fd(),
+                buff.as_mut_ptr() as *mut libc::c_void,
+                SIZE,
+                libc::MSG_DONTWAIT,
+            ) > 0
+            {}
+        }
+
+        if self.is_closed() {
+            // Wake any other sleeping ends
+            // (if none wait, it'll only leave garbage inside the pipe, but we'll close it soon
+            // anyway).
+            self.waker.wake();
+        }
+    }
+
+    /// Returns an iterator of already received signals.
+    ///
+    /// This returns an iterator over all the signal numbers of the signals received since last
+    /// time they were read (out of the set registered by this `Signals` instance). Note that
+    /// they are returned in arbitrary order and a signal number is returned only once even if
+    /// it was received multiple times.
+    ///
+    /// This method returns immediately (does not block) and may produce an empty iterator if
+    /// there are no signals ready.
+    pub fn pending(&self) -> Pending {
+        if !self.is_closed() {
+            self.flush();
+        }
+        self.waker.pending()
+    }
+
+    /// Checks the reading end of the self pipe for available signals.
+    ///
+    /// If there are no signals available or this instance was already closed it returns
+    /// [`Option::None`]. If there are some signals it returns a [`Pending`](struct.Pending)
+    /// instance wrapped inside a [`Option::Some`]. However, due to implementation details,
+    /// this still can produce an empty iterator.
+    ///
+    /// This method doesn't check the reading end itself but uses the passed in callback. This
+    /// method blocks if and only if the callback blocks trying to read some bytes.
+    pub fn wait<F>(&self, has_signals: &mut F) -> Result<Option<Pending>, Error>
```

Comment:
:thinking: After all the refactoring, is the `wait` name still appropriate?

Reply:
Yeah, `wait` isn't really meaningful anymore. So may be `poll_pending`?